content/docs/ai/ai-checkpoints.md

```diff
@@ -0,0 +1,356 @@
+**Visual Diagram: Active Branch Pattern Architecture** - Shows one-to-one project mapping, the active branch with permanent connection string, multiple snapshots on a timeline, and preview branches with temporary connection strings
+</Admonition>
+
+```text
+One Agent Project = One Neon Project:
+├── Active Branch (main/root) → postgresql://user@host/db (never changes)
+├── main (old) → backup from previous restore [can be deleted]
+├── Snapshot 1 (checkpoint at t1)
+├── Snapshot 2 (checkpoint at t2)
+└── Preview Branch (temporary, created from snapshot)
+```
+
+## Implementation
+
+### Creating checkpoints
+
+Create a snapshot to capture the current database state using the [snapshot endpoint](https://api-docs.neon.tech/reference/createsnapshot):
+
+```bash
+POST /api/v2/projects/{project_id}/branches/{branch_id}/snapshot
+```
+
+**Path parameters:**
+
+- `project_id` (string, required): The Neon project ID
+- `branch_id` (string, required): The branch ID
+
+**Query parameters:**
+
+- `lsn` (string): Target Log Sequence Number. Cannot be used with `timestamp`
+- `timestamp` (string): Target timestamp (RFC 3339). Cannot be used with `lsn`
+- `name` (string): Name for the snapshot
+- `expires_at` (string): Auto-deletion time (RFC 3339)
+
+**Example:**
+
+```bash
+curl --request POST \
+     --url 'https://console.neon.tech/api/v2/projects/{project_id}/branches/{branch_id}/snapshot?name=checkpoint-session-1&expires_at=2025-08-13T00:00:00Z' \
+     --header 'authorization: Bearer $NEON_API_KEY'
+```
+
+**When to create snapshots:**
+
+- Start of each agent session
+- Before database schema changes
+- After successful operations
+- User-initiated save points
+
+### Restoring checkpoints
+
+Restore any snapshot to the active branch using the [restore endpoint](https://api-docs.neon.tech/reference/restoresnapshot):
+
+```bash
+POST /api/v2/projects/{project_id}/snapshots/{snapshot_id}/restore
+```
+
+**Path parameters:**
+
+- `project_id` (string, required): The Neon project ID
+- `snapshot_id` (string, required): The snapshot ID
+
+**Body parameters:**
+
+- `name` (string): A name for the newly restored branch. If omitted, a default name will be generated
+- `target_branch_id` (string): The ID of the branch to restore the snapshot into. If not specified, the branch from which the snapshot was originally created will be used. **Set to your root branch ID for rollbacks to preserve connection strings**
+- `finalize_restore` (boolean): Set to `true` to finalize the restore operation immediately. This will complete the restore and move any associated computes to the new branch. Defaults to `false`. **Set to `true` when restoring to the active branch for rollbacks, `false` for preview branches**
+
+<Admonition type="warning" title="Connection Warning">
+Do not connect to the database until all operations complete. Any connection attempt before operations finish will connect to the old database state, not your restored snapshot.
```

Comment:
We should point to operations polling docs, like Andre has in his repo. This content can also be found in the docs here:
https://neon.com/docs/manage/operations#poll-operation-status
We haven't updated these docs in a while. I'd validate them against Andre's repo docs. 
Because this is critical to the flow, I wonder if we need an example directly in this doc (?)